File: backend/app/api/v1/routes/lecturer_ai.py
from __future__ import annotations

import uuid
from typing import List

from fastapi import APIRouter, Depends, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.pptx_exporter import build_slide_deck_pptx
from app.db.models.auth import User
from app.db.session import get_db
from app.dependencies.auth import require_lecturer_or_admin
from app.schemas.lecturer_ai import (
    LearningUnitItemResponse,
    LecturerSupportRequest,
    LecturerSupportResponse,
    RubricDraftRequest,
    RubricDraftResponse,
    RubricSaveRequest,
    RubricSaveResponse,
    SlideDeckExportRequest,
    SlideDeckGenerateRequest,
    SlideDeckGenerateResponse,
)
from app.services.lecturer_ai_service import LecturerAIService

router = APIRouter(prefix="/lecturers/ai", tags=["Lecturer AI"])


@router.post(
    "/support",
    status_code=status.HTTP_200_OK,
    response_model=LecturerSupportResponse,
    summary="Ask the Lecturer AI Assistant",
)
async def lecturer_support(
    body: LecturerSupportRequest,
    current_user: User = Depends(require_lecturer_or_admin),
    db: AsyncSession = Depends(get_db),
) -> LecturerSupportResponse:
    service = LecturerAIService(db)
    return await service.support(body, current_user)


# ─── Learning Units & Slide Decks ─────────────────────────────────────────────

@router.get(
    "/workspaces/{workspace_id}/learning-units",
    status_code=status.HTTP_200_OK,
    response_model=List[LearningUnitItemResponse],
    summary="Get segmented Learning Units for a teaching workspace (with auto-segmentation fallback)",
)
async def get_workspace_learning_units(
    workspace_id: uuid.UUID,
    current_user: User = Depends(require_lecturer_or_admin),
    db: AsyncSession = Depends(get_db),
) -> List[LearningUnitItemResponse]:
    service = LecturerAIService(db)
    return await service.get_workspace_learning_units(workspace_id, current_user)


@router.post(
    "/slides/{learning_unit_id}/generate",
    status_code=status.HTTP_200_OK,
    response_model=SlideDeckGenerateResponse,
    summary="Generate a structured slide deck outline from a Learning Unit",
)
async def generate_slide_deck(
    learning_unit_id: uuid.UUID,
    body: SlideDeckGenerateRequest = SlideDeckGenerateRequest(),
    current_user: User = Depends(require_lecturer_or_admin),
    db: AsyncSession = Depends(get_db),
) -> SlideDeckGenerateResponse:
    service = LecturerAIService(db)
    return await service.generate_slide_deck(
        learning_unit_id=learning_unit_id,
        current_user=current_user,
        estimated_minutes=body.estimated_minutes,
        selected_outcomes=body.selected_outcomes,
    )


@router.post(
    "/slides/export",
    status_code=status.HTTP_200_OK,
    summary="Export slide deck JSON as a .pptx presentation file",
)
async def export_slide_deck(
    body: SlideDeckExportRequest,
    current_user: User = Depends(require_lecturer_or_admin),
) -> Response:
    pptx_bytes = build_slide_deck_pptx(body.deck)
    filename = f"slide-deck-{uuid.uuid4().hex[:8]}.pptx"
    return Response(
        content=pptx_bytes,
        media_type="application/vnd.openxmlformats-officedocument.presentationml.presentation",
        headers={
            "Content-Disposition": f'attachment; filename="{filename}"',
        },
    )


# ─── Rubric Assistant ─────────────────────────────────────────────────────────

@router.post(
    "/rubrics/draft",
    status_code=status.HTTP_200_OK,
    response_model=RubricDraftResponse,
    summary="Draft or enhance rubric criteria grounded to a Question",
)
async def draft_rubric(
    body: RubricDraftRequest,
    current_user: User = Depends(require_lecturer_or_admin),
    db: AsyncSession = Depends(get_db),
) -> RubricDraftResponse:
    service = LecturerAIService(db)
    return await service.draft_rubric(body, current_user)


@router.post(
    "/rubrics/save",
    status_code=status.HTTP_200_OK,
    response_model=RubricSaveResponse,
    summary="Save rubric criteria directly to the database and attach to target Question",
)
async def save_rubric(
    body: RubricSaveRequest,
    current_user: User = Depends(require_lecturer_or_admin),
    db: AsyncSession = Depends(get_db),
) -> RubricSaveResponse:
    service = LecturerAIService(db)
    return await service.save_rubric_to_question(body, current_user)

File: frontend/lib/api/lecturer-ai.ts
import { apiClient, getToken, resolveApiUrl } from "./client";

export interface LearningUnitItem {
  id: string;
  order_index: number;
  title: string;
  summary?: string | null;
  learning_outcomes?: string[];
  start_page?: number | null;
  end_page?: number | null;
  source_material_id?: string | null;
  estimated_study_minutes: number;
  chunk_count: number;
}

export interface SlideItem {
  title: string;
  bullet_points: string[];
  visual_idea?: string | null;
  speaker_notes?: string | null;
}

export interface SlideDeckOutput {
  title: string;
  target_audience: string;
  estimated_minutes: number;
  slides: SlideItem[];
}

export interface SlideDeckGenerateResponse {
  learning_unit_id: string;
  unit_title: string;
  deck: SlideDeckOutput;
}

export interface RubricCriterionLevel {
  label: string;
  description: string;
  marks: number;
}

export interface RubricCriterion {
  title: string;
  description?: string | null;
  max_marks: number;
  order_index?: number;
  levels: RubricCriterionLevel[];
}

export interface RubricDraftOutput {
  title: string;
  description?: string | null;
  criteria: RubricCriterion[];
}

export interface RubricDraftResponse {
  question_id: string;
  rubric: RubricDraftOutput;
}

export interface RubricSaveResponse {
  question_id: string;
  rubric_id: string;
  title: string;
  criteria_count: number;
  message: string;
}

export const lecturerAiApi = {
  /**
   * Fetch ordered Learning Units for a workspace (with automatic segmentation fallback)
   */
  async getWorkspaceLearningUnits(workspaceId: string): Promise<LearningUnitItem[]> {
    return apiClient(`/lecturers/ai/workspaces/${workspaceId}/learning-units`);
  },

  /**
   * Generate an 8-15 slide deck outline from a Learning Unit
   */
  async generateSlideDeck(
    learningUnitId: string,
    estimatedMinutes: number = 45,
    selectedOutcomes?: string[]
  ): Promise<SlideDeckGenerateResponse> {
    return apiClient(`/lecturers/ai/slides/${learningUnitId}/generate`, {
      method: "POST",
      body: JSON.stringify({
        estimated_minutes: estimatedMinutes,
        selected_outcomes: selectedOutcomes,
      }),
    });
  },

  /**
   * Export a SlideDeckOutput schema to a downloadable .pptx presentation file
   */
  async exportSlideDeckPptx(deck: SlideDeckOutput): Promise<Blob> {
    const apiUrl = resolveApiUrl();
    const token = getToken();
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };
    if (token) {
      headers["Authorization"] = `Bearer ${token}`;
    }

    const res = await fetch(`${apiUrl}/lecturers/ai/slides/export`, {
      method: "POST",
      headers,
      body: JSON.stringify({ deck }),
    });

    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.message || err.detail || "Failed to export PowerPoint presentation");
    }

    return res.blob();
  },

  /**
   * Draft rubric criteria grounded to a Question entity
   */
  async draftRubric(
    questionId: string,
    totalMarks?: number,
    existingRubric?: string
  ): Promise<RubricDraftResponse> {
    return apiClient("/lecturers/ai/rubrics/draft", {
      method: "POST",
      body: JSON.stringify({
        question_id: questionId,
        total_marks: totalMarks,
        existing_rubric: existingRubric,
      }),
    });
  },

  /**
   * Save rubric directly to the database and attach to target Question
   */
  async saveRubric(
    questionId: string,
    title: string,
    criteria: RubricCriterion[],
    description?: string
  ): Promise<RubricSaveResponse> {
    return apiClient("/lecturers/ai/rubrics/save", {
      method: "POST",
      body: JSON.stringify({
        question_id: questionId,
        title,
        description,
        criteria,
      }),
    });
  },
};
